sdk/src/azure/core/az_mqtt5_connection_hfsm.c

```diff
@@ -305,30 +342,65 @@ static az_result reconnect_timeout(az_event_policy* me, az_event event)
   switch (event.type)
   {
     case AZ_HFSM_EVENT_ENTRY:
-      // TODO_L: reconnect timer (apply reconnect policy), counter not implemented.
+    {
+      int32_t random_num = 0;
+
+      _az_RETURN_IF_FAILED(az_platform_get_random(&random_num));
+      double normalized_random_num = random_num / (double)RAND_MAX;
+      int32_t random_jitter_msec = (int32_t)(
+          normalized_random_num * this_policy->_internal.options.max_random_jitter_msec);
+
+      _az_RETURN_IF_FAILED(az_platform_timer_create(
+          &this_policy->_internal.connection_timer, _connection_timeout_callback, this_policy));
+      _az_RETURN_IF_FAILED(az_platform_timer_start(
+          &this_policy->_internal.connection_timer,
+          this_policy->_internal.options.retry_delay_function(
+              this_policy->_internal.reconnect_counter == 0
+                  ? 0
+                  : this_policy->_internal.connect_time_msec,
+              this_policy->_internal.reconnect_counter,
+              this_policy->_internal.options.min_retry_delay_msec,
+              this_policy->_internal.options.max_retry_delay_msec,
+              random_jitter_msec)));
+
+      this_policy->_internal.reconnect_counter++;
+      if (this_policy->_internal.reconnect_counter
+          > this_policy->_internal.options.max_connect_attempts)
+      {
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, reconnect_timeout, faulted));
+      }
+
+      this_policy->_internal.client_certificate_index++;
+
       break;
+    }
 
     case AZ_HFSM_EVENT_EXIT:
-      // TODO_L
+      _az_RETURN_IF_FAILED(az_platform_timer_destroy(&this_policy->_internal.connection_timer));
```

Comment:
Please review some of the RETURN_IF_FAILED statements.
If the timer fails to destroy, I don't think you should be using the timer or the connection anymore. Returning the error is not sufficient in this case as it would just fail the transition.

In general, avoid return if failed in enter/ exit/ events which are meant for resource allocation/deallocation.